File: cli/src/commands/send.ts
import { Command } from 'commander'
import chalk from 'chalk'
import { NotifyClient } from '../lib/client.js'

export const sendCommand = new Command('send')
  .description('Send a notification message')
  .requiredOption('-c, --channel <type>', 'Channel type (email, sms, push)')
  .requiredOption('-t, --to <address>', 'Recipient address')
  .option('-s, --subject <subject>', 'Message subject (for email)')
  .option('-b, --body <body>', 'Message body')
  .option('--template <name>', 'Template name to use')
  .option('--var <key=value...>', 'Template variables (repeatable)', collectVars, {})
  .option('--idempotency-key <key>', 'Idempotency key to prevent duplicates')
  .option('--tags <tags...>', 'Tags (space-separated)')
  .option('--priority <n>', 'Priority 0-99 (higher = more urgent)', parseInt)
  .option('--url <url>', 'URL for click-through in push notification')
  .option('--delay <delay>', 'Delay: relative (30m, 1h, 1d, 1w) or absolute (yyyy-mm-dd hh:mm:ss)')
  .option('--attachment-name <name>', 'Attachment filename')
  .option('--attachment-url <url>', 'Attachment download URL')
  .option('--attachment-data <base64>', 'Attachment base64 data')
  .option('--format <format>', 'Body format: text, markdown, html, json', 'text')
  .option('--server <url>', 'Server URL override')
  .option('--token <token>', 'API token override')
  .action(async (opts) => {
    if (!opts.body && !opts.template) {
      console.error(chalk.red('Error: --body or --template is required'))
      process.exit(1)
    }

    if (opts.attachmentName && !opts.attachmentUrl && !opts.attachmentData) {
      console.error(chalk.red('Error: --attachment-name requires --attachment-url or --attachment-data'))
      process.exit(1)
    }

    const client = new NotifyClient(opts.server, opts.token)

    console.log(chalk.gray(`Sending ${opts.channel} notification to ${opts.to}...`))

    const result = await client.send({
      channel: opts.channel,
      to: opts.to,
      subject: opts.subject,
      body: opts.body,
      template: opts.template,
      variables: opts.var,
      idempotencyKey: opts.idempotencyKey,
      tags: opts.tags,
      priority: opts.priority,
      url: opts.url,
      delay: opts.delay,
      attachment: opts.attachmentName ? {
        name: opts.attachmentName,
        url: opts.attachmentUrl,
        data: opts.attachmentData,
      } : undefined,
      format: opts.format,
    })

    if (result.success) {
      console.log(chalk.green('✓ Message queued successfully'))
      console.log(chalk.gray(`  Message ID: ${result.data?.messageId}`))
      console.log(chalk.gray(`  Status:     ${result.data?.status}`))
    } else {
      console.error(chalk.red(`✗ Failed: ${result.error}`))
      process.exit(1)
    }
  })

function collectVars(value: string, prev: Record<string, string>) {
  const eqIdx = value.indexOf('=')
  if (eqIdx === -1) {
    console.error(chalk.red(`Invalid variable format: ${value} (expected key=value)`))
    process.exit(1)
  }
  const key = value.slice(0, eqIdx)
  const val = value.slice(eqIdx + 1)
  return { ...prev, [key]: val }
}

File: rust-server/cli/src/commands/send.rs
use clap::Args;

use crate::client::NotifyClient;
use crate::config::load_config;

#[derive(Args)]
pub struct SendArgs {
    /// Channel type (email, sms, push)
    #[arg(short, long)]
    channel: String,

    /// Recipient address
    #[arg(short, long)]
    to: String,

    /// Message subject (for email)
    #[arg(short, long)]
    subject: Option<String>,

    /// Message body
    #[arg(short, long)]
    body: Option<String>,

    /// Template name to use
    #[arg(long)]
    template: Option<String>,

    /// Template variables (key=value, repeatable)
    #[arg(long = "var", value_parser = parse_var)]
    variables: Vec<(String, String)>,

    /// Idempotency key
    #[arg(long)]
    idempotency_key: Option<String>,

    /// Tags (space-separated)
    #[arg(long, value_delimiter = ' ')]
    tags: Vec<String>,

    /// Priority 0-99 (higher = more urgent)
    #[arg(long)]
    priority: Option<u32>,

    /// URL for click-through
    #[arg(long)]
    url: Option<String>,

    /// Delay: relative (30m, 1h, 1d, 1w) or absolute (yyyy-mm-dd hh:mm:ss)
    #[arg(long)]
    delay: Option<String>,

    /// Body format: text, markdown, html, json
    #[arg(long, default_value = "text")]
    format: String,

    /// Server URL override
    #[arg(long)]
    server: Option<String>,

    /// API token override
    #[arg(long)]
    token: Option<String>,
}

fn parse_var(s: &str) -> Result<(String, String), String> {
    let (key, value) = s.split_once('=')
        .ok_or_else(|| format!("invalid format, expected key=value: {s}"))?;
    Ok((key.to_string(), value.to_string()))
}

pub async fn run(args: SendArgs) -> anyhow::Result<()> {
    if args.body.is_none() && args.template.is_none() {
        eprintln!("Error: either --body or --template is required");
        std::process::exit(1);
    }

    let config = load_config()?;
    let server = args.server.as_deref().unwrap_or(&config.server);
    let token = args.token.as_deref().unwrap_or(&config.token);

    if token.is_empty() {
        eprintln!("Error: no API token configured. Run 'notifyhub config init' first.");
        std::process::exit(1);
    }

    let client = NotifyClient::new(server, token);

    let mut msg = serde_json::json!({
        "channel": args.channel,
        "to": args.to,
        "format": args.format,
        "tags": args.tags,
    });

    if let Some(subject) = &args.subject {
        msg["subject"] = serde_json::Value::String(subject.clone());
    }
    if let Some(body) = &args.body {
        msg["body"] = serde_json::Value::String(body.clone());
    }
    if let Some(template) = &args.template {
        msg["template"] = serde_json::Value::String(template.clone());
    }
    if !args.variables.is_empty() {
        let vars: serde_json::Map<String, serde_json::Value> = args.variables.iter()
            .map(|(k, v)| (k.clone(), serde_json::Value::String(v.clone())))
            .collect();
        msg["variables"] = serde_json::Value::Object(vars);
    }
    if let Some(key) = &args.idempotency_key {
        msg["idempotencyKey"] = serde_json::Value::String(key.clone());
    }
    if let Some(priority) = args.priority {
        msg["priority"] = serde_json::Value::Number(priority.into());
    }
    if let Some(url) = &args.url {
        msg["url"] = serde_json::Value::String(url.clone());
    }
    if let Some(delay) = &args.delay {
        msg["delay"] = serde_json::Value::String(delay.clone());
    }

    let resp = client.send(&msg).await?;

    if resp.success {
        if let Some(data) = &resp.data {
            let msg_id = data.get("messageId").and_then(|v| v.as_str()).unwrap_or("unknown");
            let status = data.get("status").and_then(|v| v.as_str()).unwrap_or("unknown");
            println!("Message sent: {msg_id} (status: {status})");
        }
    } else {
        let err = resp.error.unwrap_or_else(|| "unknown error".to_string());
        eprintln!("Error: {err}");
        std::process::exit(1);
    }

    Ok(())
}
